File: app/lib/domain/models/vehicle_model.dart
import 'tipo_combustivel.dart';
import 'vehicle_type.dart';

class VehicleModel {
  final int id;
  final String marca;
  final String modelo;
  final int ano;
  final VehicleType tipo;
  final TipoCombustivel tipoCombustivel;
  final double? consumoCidadeKmL;
  final double? consumoEstradaKmL;
  final double? consumoKmPorKWh;
  final int numeroEixos;
  final double custoDesgastePorKm;

  bool get isEletrico => tipoCombustivel == TipoCombustivel.eletrico;

  /// Rótulo do passo 2 (escolha de ano/versão): "2023 · Gasolina" — um
  /// modelo flex tem 2 linhas no mesmo ano (gasolina e etanol), então o
  /// combustível precisa aparecer no rótulo pra distinguir.
  String get versaoLabel => '$ano · ${tipoCombustivel.label}';

  VehicleModel({
    required this.id,
    required this.marca,
    required this.modelo,
    required this.ano,
    required this.tipo,
    this.tipoCombustivel = TipoCombustivel.gasolina,
    this.consumoCidadeKmL,
    this.consumoEstradaKmL,
    this.consumoKmPorKWh,
    required this.numeroEixos,
    required this.custoDesgastePorKm,
  });

  String get displayName => '$marca $modelo ($ano)';

  factory VehicleModel.fromJson(Map<String, dynamic> json) {
    return VehicleModel(
      id: json['id'] as int,
      marca: json['marca'] as String,
      modelo: json['modelo'] as String,
      ano: json['ano'] as int,
      tipo: VehicleType.fromApiValue(json['tipo'] as String),
      tipoCombustivel: TipoCombustivel.fromApiValue(json['tipoCombustivel'] as String?),
      consumoCidadeKmL: (json['consumoCidadeKmL'] as num?)?.toDouble(),
      consumoEstradaKmL: (json['consumoEstradaKmL'] as num?)?.toDouble(),
      consumoKmPorKWh: (json['consumoKmPorKWh'] as num?)?.toDouble(),
      numeroEixos: json['numeroEixos'] as int,
      custoDesgastePorKm: (json['custoDesgastePorKm'] as num).toDouble(),
    );
  }
}

File: backend/src/main/java/com/rotacusto/entity/VehicleModel.java
package com.rotacusto.entity;

import com.rotacusto.entity.enums.TipoCombustivel;
import com.rotacusto.entity.enums.VehicleType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;

@Entity
@Table(name = "vehicle_models")
public class VehicleModel {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String marca;

    @Column(nullable = false)
    private String modelo;

    @Column(nullable = false)
    private Integer ano;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private VehicleType tipo;

    /**
     * Um veículo flex vira DUAS linhas de catálogo (mesmo marca/modelo/ano,
     * GASOLINA e ETANOL cada uma com seu consumo) — combustível é parte da
     * identidade do registro, não um detalhe do veículo.
     */
    @Enumerated(EnumType.STRING)
    @Column(name = "tipo_combustivel", nullable = false)
    private TipoCombustivel tipoCombustivel;

    /** Nulo quando tipoCombustivel = ELETRICO. */
    @Column(name = "consumo_cidade_km_l")
    private Double consumoCidadeKmL;

    @Column(name = "consumo_estrada_km_l")
    private Double consumoEstradaKmL;

    /** Só preenchido quando tipoCombustivel = ELETRICO. */
    @Column(name = "consumo_km_por_kwh")
    private Double consumoKmPorKWh;

    @Column(name = "numero_eixos", nullable = false)
    private Integer numeroEixos;

    @Column(name = "custo_desgaste_por_km", nullable = false)
    private Double custoDesgastePorKm;

    public VehicleModel() {
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getMarca() {
        return marca;
    }

    public void setMarca(String marca) {
        this.marca = marca;
    }

    public String getModelo() {
        return modelo;
    }

    public void setModelo(String modelo) {
        this.modelo = modelo;
    }

    public Integer getAno() {
        return ano;
    }

    public void setAno(Integer ano) {
        this.ano = ano;
    }

    public VehicleType getTipo() {
        return tipo;
    }

    public void setTipo(VehicleType tipo) {
        this.tipo = tipo;
    }

    public Double getConsumoCidadeKmL() {
        return consumoCidadeKmL;
    }

    public void setConsumoCidadeKmL(Double consumoCidadeKmL) {
        this.consumoCidadeKmL = consumoCidadeKmL;
    }

    public Double getConsumoEstradaKmL() {
        return consumoEstradaKmL;
    }

    public void setConsumoEstradaKmL(Double consumoEstradaKmL) {
        this.consumoEstradaKmL = consumoEstradaKmL;
    }

    public TipoCombustivel getTipoCombustivel() {
        return tipoCombustivel;
    }

    public void setTipoCombustivel(TipoCombustivel tipoCombustivel) {
        this.tipoCombustivel = tipoCombustivel;
    }

    public Double getConsumoKmPorKWh() {
        return consumoKmPorKWh;
    }

    public void setConsumoKmPorKWh(Double consumoKmPorKWh) {
        this.consumoKmPorKWh = consumoKmPorKWh;
    }

    public Integer getNumeroEixos() {
        return numeroEixos;
    }

    public void setNumeroEixos(Integer numeroEixos) {
        this.numeroEixos = numeroEixos;
    }

    public Double getCustoDesgastePorKm() {
        return custoDesgastePorKm;
    }

    public void setCustoDesgastePorKm(Double custoDesgastePorKm) {
        this.custoDesgastePorKm = custoDesgastePorKm;
    }
}
